Reduce flakiness in bootstrap execution

gdemonet opened the pull request:
## What

We introduce two main "waiting" methods where we already encountered failures:


#### Waiting for Salt minion to respond to the master once its keys were accepted

This was implemented as a custom runner function, `metalk8s_saltutil.wait_minions`, which takes as arguments:
- `tgt`, defaults to `'*'`, for reducing the scope of the wait (in the bootstrap script, we only wait for the bootstrap minion)
- `retry`, defaults to 10, for the amount of `test.ping` attempts before bailing out if some minion still fails to respond

#### Waiting for Kube API Server to respond after being deployed/restarted

This was implemented as a custom execution function, `metalk8s.wait_apiserver`, which takes as arguments:
- `retry`, defaults to 10, for the amount of `metalk8s_kubernetes.ping` attempts
- `interval`, defaults to 1, for the number of seconds to wait between two attempts

---

Fixes: GH-994

gdemonet committed:
Fix bootstrap script indentation

Issue: GH-994

gdemonet committed:
Clean bootstrap script

Some clean-up is done for long lines and local variables naming.
We also made sure to wait until we have a running salt-master container
before trying to exec in it.

Issue: GH-994

gdemonet committed:
Add runner function to wait for minions

When accepting a new Salt minion's key, it will not be immediately
reachable from the master, but will eventually.
This runner is thus a simple method to retry a simple `salt * test.ping`
until all minions respond.

Issue: GH-994

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/994-reduce-flakiness` into
`development/2.0`
* [w/2.1/bugfix/994-reduce-flakiness](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/994-reduce-flakiness) will be merged into `development/2.1` 
* [w/2.2/bugfix/994-reduce-flakiness](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/994-reduce-flakiness) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

gdemonet committed:
Add a module function to wait for API server

When deploying/upgrading API server, it might be unresponsive for a
short period. We introduce a simple wrapper around the `kubernetes.ping`
function to avoid race conditions.

Issue: GH-994

NicolasT commented on salt/_modules/metalk8s.py:
```diff
@@ -32,6 +32,25 @@ def deps_missing(*args, **kwargs):
+    """
+    status = False
+    attempts = 0
+    while not status and attempts < retry:
```
This busy-loops, no? Shouldn't there be some kind of wait/sleep somewhere?

gdemonet replied:
I didn't check, but from what I remember in logs, API calls with the Kubernetes client do retry and wait a little between retries. But it makes senses to wait here as well, I'll add something.

NicolasT commented on salt/_modules/metalk8s.py:
```diff
@@ -32,6 +32,25 @@ def deps_missing(*args, **kwargs):
+        attempts += 1
+
+    if not status:
+        log.error('Kubernetes apiserver failed to respond after %d attempts.',
```
No punctuation in logs.

NicolasT commented on salt/metalk8s/orchestrate/bootstrap_with_master.sls:
```diff
@@ -27,6 +27,14 @@ Bootstrap control plane:
     - pillar:
         registry_ip: {{ pillar.get('registry_ip') }}
 
+Wait for kube-apiserver to be ready:
```
Instead of a custom procedure (`metalk8s.wait_apiserver`), couldn't this be `kubernetes.ping` with a `retry`? https://docs.saltstack.com/en/latest/ref/states/requisites.html#retrying-states

gdemonet replied:
So yes, it could, but IIRC, we introduced a `cri.wait_container` procedure specifically to avoid using `retry` from Salt: rationale was that we don't like seeing error logs during the bootstrap if it's just something we are retrying.

NicolasT commented on scripts/bootstrap.sh.in:
```diff
@@ -93,7 +93,7 @@ configure_yum_local_repository() {
     gpg_keys=$(
         find "$repo_path" -maxdepth 1 -name "RPM-GPG-KEY-*" \
-	    -printf "file://%p "
+        -printf "file://%p "
```
The `-printf` is an argument to the `find`, so IMHO it makes sense to be more indented.

gdemonet replied:
Yes indeed, my bad (replaced a tab with spaces)

NicolasT commented on scripts/bootstrap.sh.in:
```diff
@@ -144,56 +144,77 @@ retry_dns_count: 3
+get_salt_container() {
+    local salt_container
+    salt_container=''
+    while [ -z "$salt_container" ]; do
```
This could loop forever, needs a break/failure at some point.

gdemonet replied:
Good point, will do.

NicolasT commented on scripts/bootstrap.sh.in:
```diff
@@ -144,56 +144,77 @@ retry_dns_count: 3
+    # TODO: Registry IP should not need a pillar override here
+    local cidr
+    local registry_ip
+    cidr=$( \
```
Could your branch be out of date?

gdemonet replied:
Probably! Will rebase.

gdemonet replied:
So I had no conflict when rebasing, did you expect this to retrieve the IP using grains, as introduced in 37340e4? I'm fine with adding it.

NicolasT replied:
I must have had another PR which isn't merged yet in mind.

NicolasT commented on scripts/bootstrap.sh.in:
```diff
@@ -144,56 +144,77 @@ retry_dns_count: 3
+    local salt_master_call
+    salt_master_call="crictl exec -i $(get_salt_container)"
+
+    # Wait for all minions to respond before proceeding
```
Why `all`? Don't we only need bootstrap? So `salt-call test.ping` should do, no?

gdemonet replied:
So:

- Yes we do only need bootstrap, I could do `salt-run metalk8s_saltutil.wait_minions tgt="$(cat /etc/salt/minion_id)"`, but I found it to be unnecessary (I'll let you choose)
- A simple `salt '*' test.ping` will always have return code 0, even if minions fail to respond, so we would have needed to parse the output of the command (I found it to be easier in Python)
- I didn't try to `salt-call test.ping` on the minion directly, maybe it waits better than `salt '*' test.ping`, and has an appropriate return code
- With a runner like this, we could include it in an orchestrate (e.g. the one used to add new nodes to the cluster), while a simple `test.ping` may be harder to use (with retries and expected response...)

NicolasT replied:
> Yes we do only need bootstrap, I could do salt-run metalk8s_saltutil.wait_minions tgt="$(cat /etc/salt/minion_id)", but I found it to be unnecessary (I'll let you choose)

I guess that's a bit of an issue in an environment where
- we need to restore the `bootstrap` node for whatever reason (i.e. re-run this script)
- for some reason, not all minions are/can be online

NicolasT replied:
Other than that, OK.

gdemonet replied:
OK, I'll add the more specific `tgt` then.

bert-e commented:
# History mismatch



Merge commit #e7e49a5018eb59a6eae477f30410ebd9ee14254c on the integration branch
`w/2.1/bugfix/994-reduce-flakiness` is merging a branch which is neither the current
branch `bugfix/994-reduce-flakiness` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `bugfix/994-reduce-flakiness` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented on salt/_runners/metalk8s_saltutil.py:
```diff
@@ -1,7 +1,12 @@
+import salt.client
 import salt.utils.extmods
 
+LOGGER = logging.getLogger(__name__)
```
Salt habits is to call this `log` (yes, unlike PEP8 dictates), but OK :)

gdemonet replied:
I see that we use this convention in other Salt modules, so I'll comply with it.

NicolasT reviewed (changes requested):
I'd like to only 'wait for what we really need' rather than be overzealous.

Other than that, LGTM if CI agrees.

gdemonet commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/994-reduce-flakiness` into
`development/2.0`
* [w/2.1/bugfix/994-reduce-flakiness](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/994-reduce-flakiness) will be merged into `development/2.1` 
* [w/2.2/bugfix/994-reduce-flakiness](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/994-reduce-flakiness) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

bert-e commented:
# History mismatch



Merge commit #ed7322eda4c78004373df0aed7c170a7f2b55be5 on the integration branch
`w/2.1/bugfix/994-reduce-flakiness` is merging a branch which is neither the current
branch `bugfix/994-reduce-flakiness` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `bugfix/994-reduce-flakiness` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/994-reduce-flakiness` into
`development/2.0`
* [w/2.1/bugfix/994-reduce-flakiness](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/994-reduce-flakiness) will be merged into `development/2.1` 
* [w/2.2/bugfix/994-reduce-flakiness](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/994-reduce-flakiness) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

gdemonet commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/994-reduce-flakiness` into
`development/2.0`
* [w/2.1/bugfix/994-reduce-flakiness](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/994-reduce-flakiness) will be merged into `development/2.1` 
* [w/2.2/bugfix/994-reduce-flakiness](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/994-reduce-flakiness) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

NicolasT commented on scripts/bootstrap.sh.in:
```diff
@@ -177,34 +191,54 @@ orchestrate_bootstrap() {
 
-    # shellcheck disable=SC2086
-    ${SALT_MASTER_CALL} salt-run state.orchestrate \
+    sync_salt_minion "$bootstrap_ip"
```
`bootstrap_id`?

gdemonet replied:
Oops

NicolasT commented on scripts/bootstrap.sh.in:
```diff
@@ -144,28 +144,42 @@ retry_dns_count: 3
+    local -r target=${1:-'*'}
+    local -r salt_master_call="crictl exec -i $(get_salt_container)"
+
+    $salt_master_call salt "$target" saltutil.sync_all refresh=True \
```
All interactions in this file targetting a minion will target the bootstrap minion, so can't you use `salt-call` on the host instead?

gdemonet replied:
Indeed, that's simpler. Will do!

bert-e commented:
# History mismatch



Merge commit #d028fbd543ff32a6c60f53bd1413b20fa55da8e2 on the integration branch
`w/2.1/bugfix/994-reduce-flakiness` is merging a branch which is neither the current
branch `bugfix/994-reduce-flakiness` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `bugfix/994-reduce-flakiness` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/994-reduce-flakiness` into
`development/2.0`
* [w/2.1/bugfix/994-reduce-flakiness](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/994-reduce-flakiness) will be merged into `development/2.1` 
* [w/2.2/bugfix/994-reduce-flakiness](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/994-reduce-flakiness) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

NicolasT reviewed (approved):
Yes please!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

gdemonet commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
